core/src/main/java/com/linecorp/armeria/client/AbstractHttpRequestHandler.java

```diff
@@ -329,6 +329,10 @@ private void fail(Throwable cause) {
     }
 
     final void failAndReset(Throwable cause) {
+        // Mark the session as unhealthy so that subsequent requests do not use it.
+        final HttpSession session = HttpSession.get(ch);
+        session.deactivate();
```

Comment:
We don't want to deactivate the session depending on the exceptions. e.g. `ResponseCompleteException`
How about deactivating it when `WriteTimeoutException` is raised?

Reply:
Agreed. Could we also rename `HttpSession.deactivate()` to `.markUnacquirable()` to make it straightforward?

Reply:
> We don't want to deactivate the session depending on the exceptions. e.g. ResponseCompleteException

Regardless of the exception type, the connection is going to reset if `failAndReset` is called. Is there any reason not to deactivate the session if `cause` is `ResponseCompleteException`.

> Could we also rename HttpSession.deactivate() to .markUnacquirable() to make it straightforward?

Sounds good.

Reply:
I realized that HTTP/2 only reset the stream. I think we need to handle it separately depending on HTTP version.